File: 930. Binary Subarrays With Sum/optimised.ts
function numSubarraysWithSum(nums: number[], goal: number): number {
	function atMost(x) {
		const len = nums.length;
		let tt = 0;
		let ans = 0;
		let i = 0;
		for (let j = 0; j < len; j++) {
			tt += nums[j];

			while (tt > x) {
				tt -= nums[i];
				i++;
			}

			ans += j - i + 1;
		}

		return ans
	}

	return atMost(goal) - atMost(goal - 1);
};

console.log(numSubarraysWithSum([1,0,1,0,1], 2))

File: 930. Binary Subarrays With Sum/optimised.py
#!/usr/bin/env python

from typing import List

def numSubarraysWithSum(nums: List[int], goal: int) -> int:
    def atMost(x):
        n = len(nums)
        ans = 0
        total = 0
        i = 0
        for j in range(n):
            total += nums[j]
            while total > x:
                total -= nums[i]
                i += 1

            ans += j - i + 1

        return ans

    return atMost(goal) - atMost(goal - 1)

print(numSubarraysWithSum([1,0,1,0,1], 2))
